File: tutorials/lua/09-hello-liveness-readiness-health/src/health_checker.lua
local json = require("dkjson")

local M = {}
local deps = {}

function M.init()
    deps = {database = "ok", cache = "ok"}
end

function M.set_dependency(name, status)
    deps[name] = status
end

function M.check_all()
    local copy = {}
    for k, v in pairs(deps) do copy[k] = v end
    return copy
end

function M.status()
    for _, v in pairs(deps) do
        if v ~= "ok" then return "degraded" end
    end
    return "ok"
end

function M.liveness_json()
    return json.encode({status = "ok"})
end

function M.readiness_json()
    local s = M.status()
    return json.encode({status = s}), s == "ok" and 200 or 503
end

function M.health_json()
    local checks = {}
    for name, status in pairs(deps) do
        checks[name] = {status = status}
    end
    return json.encode({status = M.status(), checks = checks})
end

return M

File: tutorials/erlang/09-hello-liveness-readiness-health/src/health_checker.erl
-module(health_checker).
-behaviour(gen_server).
-export([start_link/0, check_all/0, status/0, set_dependency/2]).
-export([init/1, handle_call/3, handle_cast/2]).

start_link() -> gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

check_all() -> gen_server:call(?MODULE, check_all).
status() -> gen_server:call(?MODULE, status).
set_dependency(Name, Status) -> gen_server:call(?MODULE, {set_dep, Name, Status}).

init([]) -> {ok, #{database => ok, cache => ok}}.

handle_call(check_all, _From, State) -> {reply, State, State};
handle_call(status, _From, State) ->
    S = case lists:all(fun(V) -> V =:= ok end, maps:values(State)) of
        true -> ok;
        false -> degraded
    end,
    {reply, S, State};
handle_call({set_dep, Name, Status}, _From, State) ->
    {reply, ok, State#{Name => Status}}.

handle_cast(_Msg, State) -> {noreply, State}.
